algorithms/finetune/cal_ql.py

```diff
@@ -0,0 +1,1209 @@
+        _, log_probs = self.tanh_gaussian(mean, log_std, False)
+        return log_probs
+
+    def forward(
+        self,
+        observations: torch.Tensor,
+        deterministic: bool = False,
+        repeat: bool = None,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        if repeat is not None:
+            observations = extend_and_repeat(observations, 1, repeat)
+        base_network_output = self.base_network(observations)
+        mean, log_std = torch.split(base_network_output, self.action_dim, dim=-1)
+        log_std = self.log_std_multiplier() * log_std + self.log_std_offset()
+        actions, log_probs = self.tanh_gaussian(mean, log_std, deterministic)
+        return self.max_action * actions, log_probs
+
+    @torch.no_grad()
+    def act(self, state: np.ndarray, device: str = "cpu"):
+        state = torch.tensor(state.reshape(1, -1), device=device, dtype=torch.float32)
+        with torch.no_grad():
+            actions, _ = self(state, not self.training)
+        return actions.cpu().data.numpy().flatten()
+
+
+class FullyConnectedQFunction(nn.Module):
+    def __init__(
+        self,
+        observation_dim: int,
+        action_dim: int,
+        orthogonal_init: bool = False,
+        n_hidden_layers: int = 2,
+    ):
+        super().__init__()
+        self.observation_dim = observation_dim
+        self.action_dim = action_dim
+        self.orthogonal_init = orthogonal_init
+
+        layers = [
+            nn.Linear(observation_dim + action_dim, 256),
+            nn.ReLU(),
+        ]
+        for _ in range(n_hidden_layers - 1):
+            layers.append(nn.Linear(256, 256))
+            layers.append(nn.ReLU())
+        layers.append(nn.Linear(256, 1))
+
+        self.network = nn.Sequential(*layers)
+        if orthogonal_init:
+            self.network.apply(lambda m: init_module_weights(m, True))
+        else:
+            init_module_weights(self.network[-1], False)
+
+    def forward(self, observations: torch.Tensor, actions: torch.Tensor) -> torch.Tensor:
+        multiple_actions = False
+        batch_size = observations.shape[0]
+        if actions.ndim == 3 and observations.ndim == 2:
+            multiple_actions = True
+            observations = extend_and_repeat(observations, 1, actions.shape[1]).reshape(
+                -1, observations.shape[-1]
+            )
+            actions = actions.reshape(-1, actions.shape[-1])
+        input_tensor = torch.cat([observations, actions], dim=-1)
+        q_values = torch.squeeze(self.network(input_tensor), dim=-1)
+        if multiple_actions:
+            q_values = q_values.reshape(batch_size, -1)
+        return q_values
+
+
+class Scalar(nn.Module):
+    def __init__(self, init_value: float):
+        super().__init__()
+        self.constant = nn.Parameter(torch.tensor(init_value, dtype=torch.float32))
+
+    def forward(self) -> nn.Parameter:
+        return self.constant
+
+
+class CalQL:
+    def __init__(
```

Comment:
typings...

Reply:
Before we decided not to put typings for model and optimizers